File: src/http/server/admin/usage.rs
//! Read-only admin observability endpoints (§15, native-only): usage rows,
//! usage rollups, persisted credential health, and the request logs. Plus one
//! LIVE endpoint — `credential_usage` queries the upstream usage API on demand.
//!
//! The read-only handlers serialize their records directly — none of
//! `Usage` / `UsageRollup` / `CredentialStatus` / `DownstreamRequest` /
//! `UpstreamRequest` carries a secret. The log records' `headers_json` / `body`
//! are §14.3-redacted at capture, so reading them back is safe. Mounted behind
//! `require_admin`.

use axum::Json;
use axum::extract::{Path, Query, State};
use serde::Deserialize;

use super::crud::internal;
use crate::api::error::ApiError;
use crate::api::routing::EffectiveRoute;
use crate::app::AppState;
use crate::channel::UsageSnapshot;
use crate::store::persistence::UsageQuery as StoreUsageQuery;
use crate::store::persistence::records::{
    AuditLog, CredentialStatus, DownstreamRequest, UpstreamRequest, Usage, UsageRollup,
};

/// `?limit=N` for the audit listing; defaults to 100, capped at 1000.
#[derive(Debug, Clone, Copy, Deserialize)]
pub struct UsageQuery {
    pub limit: Option<u64>,
}

const DEFAULT_USAGE_LIMIT: u64 = 100;
const MAX_USAGE_LIMIT: u64 = 1000;

/// The usage-explorer selector (B4): optional filters + `before_id` keyset
/// cursor + `limit` (default 100, capped 1000). All filter params are optional;
/// omitting them yields the recency-only listing.
#[derive(Debug, Clone, Deserialize)]
pub struct UsageFilterQuery {
    pub at_from: Option<i64>,
    pub at_to: Option<i64>,
    pub provider_id: Option<i64>,
    pub user_id: Option<i64>,
    pub route_name: Option<String>,
    pub model: Option<String>,
    pub before_id: Option<i64>,
    pub limit: Option<u64>,
}

/// `GET /admin/usage` — filtered + keyset-paginated usage rows (id desc). With
/// no params this is the recency-only listing; `before_id` pages backwards.
pub async fn list_usage(
    State(state): State<AppState>,
    Query(q): Query<UsageFilterQuery>,
) -> Result<Json<Vec<Usage>>, ApiError> {
    let limit = q.limit.unwrap_or(DEFAULT_USAGE_LIMIT).min(MAX_USAGE_LIMIT);
    let query = StoreUsageQuery {
        at_from: q.at_from,
        at_to: q.at_to,
        provider_id: q.provider_id,
        user_id: q.user_id,
        route_name: q.route_name,
        model: q.model,
        before_id: q.before_id,
        limit,
    };
    Ok(Json(
        state
            .persistence
            .query_usages(&query)
            .await
            .map_err(internal)?,
    ))
}

/// The `?granularity=&from=&to=` selector for rollups.
#[derive(Debug, Clone, Deserialize)]
pub struct RollupQuery {
    pub granularity: String,
    pub from: i64,
    pub to: i64,
}

/// `GET /admin/usage-rollups?granularity=hour|day|week|month&from=&to=` —
/// rollup buckets for one granularity in `[from, to]`.
pub async fn list_usage_rollups(
    State(state): State<AppState>,
    Query(q): Query<RollupQuery>,
) -> Result<Json<Vec<UsageRollup>>, ApiError> {
    if !matches!(q.granularity.as_str(), "hour" | "day" | "week" | "month") {
        return Err(ApiError::BadRequest(
            "granularity must be one of hour|day|week|month".into(),
        ));
    }
    Ok(Json(
        state
            .persistence
            .list_usage_rollups(&q.granularity, q.from, q.to, None)
            .await
            .map_err(internal)?,
    ))
}

/// `GET /admin/audit?limit=N` — the most recent audit rows (id desc). Audit
/// rows carry only method/path/actor/status/ip; never a secret.
pub async fn list_audit(
    State(state): State<AppState>,
    Query(q): Query<UsageQuery>,
) -> Result<Json<Vec<AuditLog>>, ApiError> {
    let limit = q.limit.unwrap_or(DEFAULT_USAGE_LIMIT).min(MAX_USAGE_LIMIT);
    Ok(Json(
        state
            .persistence
            .list_audit_logs(limit)
            .await
            .map_err(internal)?,
    ))
}

/// `GET /admin/credentials/{id}/status` — the persisted credential health
/// snapshots (§16.3 edge-triggered breaker/cooldown rows) for one credential.
pub async fn credential_status(
    State(state): State<AppState>,
    Path(id): Path<i64>,
) -> Result<Json<Vec<CredentialStatus>>, ApiError> {
    Ok(Json(
        state
            .persistence
            .list_credential_statuses(id)
            .await
            .map_err(internal)?,
    ))
}

/// `GET /admin/credential-statuses` — all persisted credential health snapshots
/// (B5 batch endpoint). Callers group by `credential_id`; at most one row per
/// `(credential_id, channel)` pair is stored.
pub async fn credential_statuses(
    State(state): State<AppState>,
) -> Result<Json<Vec<CredentialStatus>>, ApiError> {
    Ok(Json(
        state
            .persistence
            .list_all_credential_statuses()
            .await
            .map_err(internal)?,
    ))
}

/// `GET /admin/credentials/{id}/usage` — LIVE per-credential upstream usage /
/// quota for OAuth subscription channels. Resolves the credential's client,
/// refreshes the token if stale, and queries the provider's usage endpoint on
/// demand. NOT cached — call sparingly (some upstreams rate-limit it). Channels
/// without a usage endpoint (api-key / vertex) return 400.
pub async fn credential_usage(
    State(state): State<AppState>,
    Path(id): Path<i64>,
) -> Result<Json<UsageSnapshot>, ApiError> {
    crate::credentials::usage::fetch_usage(&state, id)
        .await
        .map(Json)
        .map_err(usage_error)
}

/// Map a [`UsageError`](crate::credentials::usage::UsageError) onto the admin
/// API status set: missing entities → 404, no-endpoint / bad credential → 400,
/// transport / decrypt / non-2xx upstream → 500.
fn usage_error(e: crate::credentials::usage::UsageError) -> ApiError {
    use crate::credentials::usage::UsageError as U;
    match e {
        U::CredentialNotFound | U::ProviderNotFound | U::UnknownChannel(_) => {
            ApiError::NotFound(e.to_string())
        }
        U::Unsupported | U::Channel(_) => ApiError::BadRequest(e.to_string()),
        U::Decrypt(_) | U::Upstream(_) | U::Status(_) => ApiError::Internal(e.to_string()),
    }
}

/// `GET /admin/providers/{provider_id}/routing-rules/effective` — computed
/// routing decision matrix (3 ops × 4 content-gen kinds) for one provider.
/// Shows DEFAULT decisions (from [`crate::transform::routing::decide`]) with
/// each cell tagged `source: "default"` or `"override"` based on stored rules.
pub async fn effective_routing(
    State(state): State<AppState>,
    Path(provider_id): Path<i64>,
) -> Result<Json<Vec<EffectiveRoute>>, ApiError> {
    crate::api::routing::effective_routes(&state, provider_id)
        .await
        .map(Json)
}

/// `GET /admin/logs/{request_id}/downstream` — downstream (client → proxy) log
/// entries correlated by `request_id` (§15).
pub async fn downstream_logs(
    State(state): State<AppState>,
    Path(request_id): Path<String>,
) -> Result<Json<Vec<DownstreamRequest>>, ApiError> {
    Ok(Json(
        state
            .persistence
            .list_downstream_requests(&request_id)
            .await
            .map_err(internal)?,
    ))
}

/// `GET /admin/logs/{request_id}/upstream` — upstream (proxy → provider) log
/// entries correlated by `request_id` (§15).
pub async fn upstream_logs(
    State(state): State<AppState>,
    Path(request_id): Path<String>,
) -> Result<Json<Vec<UpstreamRequest>>, ApiError> {
    Ok(Json(
        state
            .persistence
            .list_upstream_requests(&request_id)
            .await
            .map_err(internal)?,
    ))
}

File: console/src/api/usage.ts
import { infiniteQueryOptions, queryOptions } from "@tanstack/react-query";
import { api } from "./http";

export interface Usage {
  id: number;
  request_id: string;
  at: number;
  route_name: string | null;
  provider_id: number | null;
  credential_id: number | null;
  org_id: number | null;
  team_id: number | null;
  user_id: number | null;
  user_key_id: number | null;
  operation: string;
  kind: string;
  model: string | null;
  input_tokens: number;
  output_tokens: number;
  cache_read_tokens: number;
  cache_creation_5m_tokens: number;
  cache_creation_1h_tokens: number;
  cost: string;
  latency_ms: number;
  usage_source: string;
  ended: string;
}

export interface UsageRollup {
  id: number;
  granularity: string;
  bucket_start: number;
  provider_id: number | null;
  org_id: number | null;
  team_id: number | null;
  user_id: number | null;
  route_name: string | null;
  model: string | null;
  requests: number;
  input_tokens: number;
  output_tokens: number;
  cost: string;
}

export interface DownstreamRequest {
  id: number;
  request_id: string;
  at: number;
  method: string;
  path: string;
  query: string | null;
  status: number;
  headers_json: unknown;
  body: string | null;
  response_body: string | null;
}

export interface UpstreamRequest {
  id: number;
  request_id: string;
  at: number;
  provider_id: number | null;
  credential_id: number | null;
  url: string;
  method: string;
  status: number;
  latency_ms: number;
  headers_json: unknown;
  body: string | null;
  response_body: string | null;
}

export interface AuditLog {
  id: number;
  at: number;
  actor_id: number | null;
  actor_name: string | null;
  action: string;
  target: string;
  status: number;
  source_ip: string | null;
}

export interface CredentialStatus {
  id: number;
  credential_id: number;
  channel: string;
  health_kind: string;
  health_json: { state?: string; open_until?: number; reason?: string } | null;
  checked_at: number | null;
  last_error: string | null;
}

export interface UsageFilter {
  at_from?: number;
  at_to?: number;
  provider_id?: number;
  user_id?: number;
  route_name?: string;
  model?: string;
  before_id?: number;
  limit?: number;
}

function usageQs(f: UsageFilter): string {
  const p = new URLSearchParams();
  if (f.limit != null) p.set("limit", String(f.limit));
  if (f.before_id != null) p.set("before_id", String(f.before_id));
  if (f.at_from != null) p.set("at_from", String(f.at_from));
  if (f.at_to != null) p.set("at_to", String(f.at_to));
  if (f.provider_id != null) p.set("provider_id", String(f.provider_id));
  if (f.user_id != null) p.set("user_id", String(f.user_id));
  if (f.route_name) p.set("route_name", f.route_name);
  if (f.model) p.set("model", f.model);
  const s = p.toString();
  return s ? `?${s}` : "";
}

export const usageQuery = (f: UsageFilter) =>
  queryOptions({
    queryKey: ["usage", f],
    queryFn: () => api<Usage[]>(`/admin/usage${usageQs(f)}`),
  });

export const PAGE = 50;

// Infinite (keyset) variant for the usage explorer.
// Each page is Usage[] DESC by id; cursor = last row id; undefined when page underfills.
export const usageInfiniteQuery = (f: Omit<UsageFilter, "before_id" | "limit">) =>
  infiniteQueryOptions({
    queryKey: ["usage", "infinite", f],
    queryFn: ({ pageParam }) =>
      api<Usage[]>(`/admin/usage${usageQs({ ...f, before_id: pageParam ?? undefined, limit: PAGE })}`),
    initialPageParam: undefined as number | undefined,
    getNextPageParam: (last: Usage[]) =>
      last.length >= PAGE ? last[last.length - 1].id : undefined,
  });

/** Recent downstream request logs (id desc, keyset-paginated). */
export const logsInfiniteQuery = () =>
  infiniteQueryOptions({
    queryKey: ["logs", "infinite"],
    queryFn: ({ pageParam }) =>
      api<DownstreamRequest[]>(`/admin/logs?limit=${PAGE}${pageParam != null ? `&before_id=${pageParam}` : ""}`),
    initialPageParam: undefined as number | undefined,
    getNextPageParam: (last: DownstreamRequest[]) =>
      last.length >= PAGE ? last[last.length - 1].id : undefined,
  });

export const rollupsQuery = (granularity: string, from: number, to: number) =>
  queryOptions({
    queryKey: ["usage-rollups", granularity, from, to],
    queryFn: () =>
      api<UsageRollup[]>(
        `/admin/usage-rollups?granularity=${granularity}&from=${from}&to=${to}`,
      ),
  });

export const downstreamLogsQuery = (rid: string) =>
  queryOptions({
    queryKey: ["logs", rid, "downstream"],
    queryFn: () => api<DownstreamRequest[]>(`/admin/logs/${rid}/downstream`),
  });

export const upstreamLogsQuery = (rid: string) =>
  queryOptions({
    queryKey: ["logs", rid, "upstream"],
    queryFn: () => api<UpstreamRequest[]>(`/admin/logs/${rid}/upstream`),
  });

export const auditQuery = (limit = 100) =>
  queryOptions({
    queryKey: ["audit", limit],
    queryFn: () => api<AuditLog[]>(`/admin/audit?limit=${limit}`),
  });

export const credentialStatusesQuery = queryOptions({
  queryKey: ["credential-statuses"],
  queryFn: () => api<CredentialStatus[]>("/admin/credential-statuses"),
  staleTime: 30_000,
});
